lxd/migrate_container.go

```diff
@@ -696,14 +697,16 @@ func (s *migrationSourceWs) Do(migrateOp *operations.Operation) error {
 		}
 	}
 
-	if s.live || (header.Criu != nil && *header.Criu == migration.CRIUType_NONE) {
-		err = driver.SendAfterCheckpoint(s.fsConn, bwlimit)
-		if err != nil {
-			return abort(err)
+	if pool == nil {
+		if s.live || (respHeader.Criu != nil && *respHeader.Criu == migration.CRIUType_NONE) {
```

Comment:
@stgraber I have a question about this part as not familiar with CRIU. From what I can tell of the existing code, if a live migration is requested (s.live will be true) then CRIU is invoked via ct.Migrate() to transfer state of container to target. 

If not doing a live migration (s.live is false) then this implies that the container is already stopped and thus we do not need to do a final storage transfer via `legacyDriver.SendAfterCheckpoint`. 

However the bit that compares the respHeader.Criu to migration.CRIUType_NONE has me confused.
Why would CRIU be used if a live migration was occurring? Why would this need a second sync of the storage layer?

Thanks

Reply:
The workflow with CRIU usually is:
 - Filesystem transfer while instance is running (partial & inconsistent)
 - Checkpoint of the source
 - Filesystem transfer of whatever changed since last run (now consistent)

As for the weird check, looking around, I think the explanation is that `s.live` indicates whether we're doing a stateful migration using CRIU using the process above. But we can also do a stateless migration, of a running container which doesn't involve CRIU, that's what the second case covers.

That's confusing because it's abusing the CRIU fields in the migration struct for this, but basically if `s.live` is false and `Criu` is set to CRIUTYPE_NONE rather than `nil`, it indicates that the source container was running and that we'll do a two stage transfer to minimize downtime.

Reply:
@stgraber following on from our convo yesterday regarding the stop/start of a container during the 2nd phase of the filesystem sync. I am pretty confident now that the reason I can't see this step in the code is because its not happening.

I've setup a cluster using LXD 3.18 (so before any of my storage migration changes had been merged) using kernel 4.15 without criu installed. 

The nodes are called ubuntu01, ubuntu02, ubuntu03 all using a dir pool.

I created a container, c1, and launched it on ubuntu01.

Then I tried to migrate it whilst running:

```
lxc move c1 c2 --target ubuntu02
Error: Migration API failure: Container is running
```

Then I tried to migrate it whilst running using --stateless:

```
lxc move c1 c2 --target ubuntu02 --stateless
Error: The --stateless flag can't be used with --target
```

I also tried just renaming it:

```
lxc move c1 c2 
Error: Renaming of running container not allowed
```

So far I haven't been able to trigger the migration logic on a running container, so next I tried just doing a copy rather than a move (thinking that a copy should still copy the container in a consistent manner).

First running without --stateless, as expected fails due to no CRIU:

```
lxc copy c1 c2 --target ubuntu02  --stateless
Error: Unable to perform container live migration. CRIU isn't installed on the source server
```

Now with the --stateless flag:

Before running copy, I checked the latest log inside the container:

```
lxc exec c1 -- journalctl  -n1
-- Logs begin at Wed 2019-12-04 09:27:17 UTC, end at Wed 2019-12-04 09:31:55 UTC. --
Dec 04 09:31:55 c1 systemd[1]: Started Message of the Day.
```


```
lxc copy c1 c2 --target ubuntu02 --stateless
```

This succeeds, but I can see no messages in the lxd log suggesting the container has been stopped/started, and inside the container's log no new messages:

```
user@ubuntu03:~$ lxc exec c1 -- journalctl  -n1
-- Logs begin at Wed 2019-12-04 09:27:17 UTC, end at Wed 2019-12-04 09:31:55 UTC. --
Dec 04 09:31:55 c1 systemd[1]: Started Message of the Day.
```

If I've missed something please can you help me see a scenario where the container is stopped/started between the first and second file syncs?

Thanks

Reply:
Don't test in a cluster, clustering doesn't support live migration at all and uses a slightly different, weird path to handle internal moves.

Reply:
`lxc move blah remote: --stateless` when `blah` is running is what should hit this code path.

Reply:
@stgraber thanks, have tried that approach now (and did a similar lxc copy), and same effect, the `SendAfterCheckpoint` function is called during a `--stateless` transfer, but it is executed before the container is stopped, and I cannot find any indication that it is run again after the container is stopped (indeed when doing a copy rather than move, the container isnt stopped).

When doing a move, the container is stopped and deleted, but not before `SendAfterCheckpoint` is run. So it looks like a consistent running copy/move is only possible using criu (as there is a good comment around the use of an "action script" written by `writeActionScript` that is used to prevent criu from reactivating the container until after the migration operation has finished).

Reply:
Ok, thanks for looking into it. We'll probably need to sort this out at some point.
A gap we have in current API is that we can't tell if we're copying something as part of a move or as part of a copy.

If copying, it's fine not to stop the source, though freezing would be nice to do. In a move case, it'd be better to stop the source between the first and second rsync, then delete once successfully transferred.

Anyway, not something to worry about now.